### ISSUE TITLE
Nix: "error while loading shared libraries: libHSterminfo-0.4.0.2-ghc8.0.2.so" after Nix garbage collection

### ISSUE THREAD

**Issue by fmthoma** [commercialhaskell/intero#407]
Over the weekend, I updated NixOS (16.09 -> 17.03). Today, I opened Spacemacs and found intero not working any more. The issue can be reproduced on the command line:

```
> stack exec intero
/home/fthoma/.stack/snapshots/x86_64-linux-nix/lts-8.4/8.0.2/bin/intero: error while loading shared libraries: libHSterminfo-0.4.0.2-ghc8.0.2.so: cannot open shared object file: No such file or directory
```

Intero version (according to the install message in Spacemacs): 0.1.20
Stack version: 1.3.2 x86_64
LTS: 8.4
Nix version: 1.11.8

```
> cat ~/.stack/config.yaml
nix:
  enable: true
```

**Comment by fmthoma** [comment]
Some more info:

```
> stack exec -- which ghc
/nix/store/hg0ppf88q1rw0q7x3zqhlxbnviq0ddcg-ghc-8.0.2/bin/ghc
> ls -l /nix/store/hg0ppf88q1rw0q7x3zqhlxbnviq0ddcg-ghc-8.0.2/lib/ghc-8.0.2/terminfo-0.4.0.2
total 2440
-r--r--r-- 1 root root 1008278 Jan  1  1970 libHSterminfo-0.4.0.2.a
-r-xr-xr-x 1 root root  233240 Jan  1  1970 libHSterminfo-0.4.0.2-ghc8.0.2.so*
-r--r--r-- 1 root root 1245768 Jan  1  1970 libHSterminfo-0.4.0.2_p.a
dr-xr-xr-x 3 root root    4096 Jan  1  1970 System/
```
which means that the module exists, in principle. The corresponding Nix package is also referenced in the environment:
```
> stack exec env | grep hg0ppf88q1rw0q7x3zqhlxbnviq0ddcg | cut -d= -f1
GHC_PACKAGE_PATH
LD_LIBRARY_PATH
PATH
STACK_IN_NIX_EXTRA_ARGS
_PATH
nativeBuildInputs
```
However, it is not recognized by intero.

**Comment by fmthoma** [comment]
Ouch. I have found the culprit: `intero` seems to be statically (?) linked to the `libHSterminfo-0.4.0.2-ghc8.0.2.so` it was compiled with. Now, Nix will re-install programs (such as `ghc`) quite liberally whenever there is an update to the package in `nixpkgs`, so there are several installations of `ghc-8.0.2` on my machine, and in particular, the ghc installation invoked by stack is not the same as the one to which intero is linked.

Now I did a Nix garbage collection last weekend, removing all »dead« packages from the Nix store. One of these removed packages was the `ghc-8.0.2` derivation intero was linked to.

Quick solution was to delete the entire LTS (`~/.stack/snapshots/x86_64-linux-nix/lts-8.4`), rebuild the project (and all its dependencies) and re-install intero (by restarting Spacemacs).

If you agree, I'll keep this issue open since the underlying problem (linking to an absolute path of `libHSterminfo-0.4.0.2-ghc8.0.2.so` rather than looking it up on the environment) still persists (and may bite other Nix users as well!).
